### PR TITLE
NIAD-757 Get latest SCR ID

### PR DESCRIPTION
**Pull request by dawidskainos**
## Summary
* Routine Change
* :exclamation: Breaking Change
* :robot: Operational or Infrastructure Change
* :sparkles: New Feature
* :warning: Potential issues that might be caused by this change

Add any other relevant notes or explanations here. **Remove this line if you have nothing to add.**

## Reviews Required
* [x] Dev
* [ ] Test
* [ ] Tech Author
* [ ] Product Owner

## Review Checklist
:information_source: This section is to be filled in by the **reviewer**.

* [ ] I have reviewed the changes in this PR and they fill all or part of the acceptance criteria of the ticket, and the code is in a mergeable state.
* [ ] If there were infrastructure, operational, or build changes, I have made sure there is sufficient evidence that the changes will work.
* [ ] I have ensured the changelog has been updated by the submitter, if necessary.

### CODE REVIEW

File: docker/service/src/main/java/uk/nhs/adaptors/scr/exceptions/NHSCodings.java [hunk 1]
```diff
@@ -1,60 +0,0 @@
-package uk.nhs.adaptors.scr.exceptions;
-
-import org.hl7.fhir.r4.model.Coding;
-
-public enum NHSCodings {
```
Comment (bartek-sarul) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/exceptions/NHSCodings.java:5]: Why is this no longer needed/used?

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/fhir/GetScrController.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+
+    @GetMapping(path = "/DocumentReference",
+        consumes = {APPLICATION_FHIR_JSON_VALUE},
+        produces = {APPLICATION_FHIR_JSON_VALUE})
+    @SuppressWarnings("checkstyle:parameternumber")
+    public String getScrId(@RequestHeader("Nhsd-Asid") @NotNull String nhsdAsid,
+                           @RequestHeader("client-ip") @NotNull String clientIp,
+                           @RequestHeader("client-request-url") @NotNull String clientRequestUrl,
+                           @RequestParam String patient,
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/fhir/GetScrController.java:44]: Good point

File: docker/service/src/main/java/uk/nhs/adaptors/scr/exceptions/RequestValidationException.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package uk.nhs.adaptors.scr.exceptions;
+
+public class RequestValidationException extends BadRequestException {
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/exceptions/RequestValidationException.java:3]: Will change to BadRequestException

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/fhir/GetScrController.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                           @RequestParam(required = false) String type,
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/fhir/GetScrController.java:45]: Expected type is http://snomed.info/sct|196981000000101 so enum will not help much

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/fhir/GetScrController.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                           @RequestParam(name = "_sort", required = false) String sort,
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/fhir/GetScrController.java:46]: Expected value is lowercase, so in order to make enum work we would need to register Converters in Spring - which does not sound like a save of time/code. Or is it simpler?

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/fhir/GetScrController.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+    private String checkAndExtractNhsNumber(String patientId) {
+        if (isNotEmpty(patientId) && patientId.startsWith(PATIENT_ID_PREFIX)) {
+            String nhsNumber = patientId.replace(PATIENT_ID_PREFIX, "");
+            if (isNotEmpty(nhsNumber)) {
+                return nhsNumber;
+            }
+        }
+
+        throw new RequestValidationException(String.format("Invalid value - %s in field 'patient'", patientId));
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/fhir/GetScrController.java:74]: Message was taken from: https://nhsd-confluence.digital.nhs.uk/display/APM/Spine+Endpoint+details

File: docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/fhir/GetScrController.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        checkGetScrCountParam(count);
+
+        Bundle bundle = getScrService.getScrId(nhsNumber, nhsdAsid, clientIp, baseUrl);
+
+        return fhirParser.encodeResource(APPLICATION_FHIR_JSON, bundle);
+    }
+
+    private String extractBaseUrl(String clientRequestUrl, String requestUri) {
+        int uriIndexOf = clientRequestUrl.indexOf(requestUri);
+        return uriIndexOf >= 0 ? clientRequestUrl.substring(0, uriIndexOf) : clientRequestUrl;
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/controllers/fhir/GetScrController.java:63]: RequestUri does not contain query params

File: docker/service/src/main/java/uk/nhs/adaptors/scr/clients/SpineClient.java [hunk 8]
```diff
@@ -98,6 +101,27 @@ public ProcessingResult getScrProcessingResult(String contentLocation, long init
         });
     }
 
+    @SneakyThrows
+    public SpineHttpClient.Response sendGetScrId(String requestBody, String nhsdAsid) {
+        LOGGER.debug("Sending GET SCR ID request to SPINE: {}", requestBody);
+        var request = new HttpPost(spineConfiguration.getUrl() + spineConfiguration.getPsisQueriesEndpoint());
+        request.addHeader("SOAPAction", "urn:nhs:names:services:psisquery/QUPC_IN180000SM04");
+        request.addHeader("Content-Type", TEXT_XML_VALUE);
+        request.addHeader("nhsd-asid", nhsdAsid);
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/clients/SpineClient.java:110]: Good point

File: docker/service/src/main/java/uk/nhs/adaptors/scr/exceptions/NHSCodings.java (hunk 1 above)
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/exceptions/NHSCodings.java:5]: This has changed due to alignment to ACS setPermission error response designed by IOPS. See at bottom: https://simplifier.net/guide/SummaryCareRecordWithCodedData/SetPermission

**Review by bartek-sarul — APPROVED** [review]

**Pull request merged**